README.md
# Bolacha

Bolacha is a simple http client that allows you testing http requests,
handling cookies and file upload.

# Why "Bolacha" ?

"Bolacha" means "cookie" in Brazillian Portuguese.

Since one of the killer features of Bolacha is dealing with cookies
and session for you, I took very fair to name the project as "Bolacha".

# Installation

     pip install bolacha >= 0.6.0

# BASIC USAGE

Will authenticate, and make a upload:

    from bolacha import Bolacha

    b = Bolacha()

    login_data = {'username': 'admin', 'password': 'qwerty'}

    b.request('http://my-website.com/login', 'POST', body=login_data)

    data = {
        'title': 'Some pictures of my vacations at Rio de Janeiro',
        'year': '2009',
        'pictures': [
            open('/home/some_user/sugar-loaf.jpg'),
            open('/home/some_user/christ-redeemer.jpg'),
        ]
    }
    b.post('http://my-website.com/upload', body=data)

    headers, body = b.get('http://my-website.com/pictures/2009')

    assert 'vacations at Rio de Janeiro' in body
    assert 'text/html' in headers['Content-Type']


debian/changelog
python-bolacha (0.2.1) experimental; urgency=low

  * New release, a big bug fix.

 -- Gabriel Falcão Gonçalves de Moura <gabriel.falcao@corp.globo.com>  Mon, 10 Aug 2009 16:05:50 -0300

python-bolacha (0.2) experimental; urgency=low

  * New release!

 -- Gabriel Falcão Gonçalves de Moura <gabriel@nacaolivre.org>  Sun, 09 Aug 2009 00:28:05 -0300

python-bolacha (0.1) experimental; urgency=low

  * Initial release.

 -- Gabriel Falcão Gonçalves de Moura <gabriel@nacaolivre.org>  Sun, 09 Aug 2009 00:16:51 -0300


tests/unit/utils.py
#!/usr/bin/env python
# -*- coding: utf-8; -*-
#
# Copyright (C) 2009 Gabriel Falcão <gabriel@nacaolivre.org>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public
# License along with this program; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.

import re
import sys

def assert_raises(exception, callable, *args, **kwargs):
    '''
    Discussion
       assert_raises() adds two optional arguments: "exc_args"
       and "exc_pattern". "exc_args" is a tuple that is expected
       to match the .args attribute of the raised exception.
       "exc_pattern" is a compiled regular expression that the
       stringified raised exception is expected to match.

    Original url: http://code.activestate.com/recipes/307970/
    Author: Trent Mick

    Usage: assert_raises(ExceptionType, method_to_execute,
                          arguments_to_method, kwargs_to_method,
                          exc_pattern=r'^.+$')
    Please note that exc_pattern is not required, but if passed
    matches the exception message.

    Fail Conditions
    Fails on exception not raised, wrong exception type or
    invalid exception message.
    '''

    if "exc_args" in kwargs:
        exc_args = kwargs["exc_args"]
        del kwargs["exc_args"]
    else:
        exc_args = None
    if "exc_pattern" in kwargs:
        exc_pattern = kwargs["exc_pattern"]
        if isinstance(exc_pattern, basestring):
            exc_pattern = re.compile(exc_pattern)

        del kwargs["exc_pattern"]
    else:
        exc_pattern = None

    argv = [repr(a) for a in args]\
           + ["%s=%r" % (k,v)  for k,v in kwargs.items()]
    callsig = "%s(%s)" % (callable.__name__, ", ".join(argv))

    try:
        callable(*args, **kwargs)
    except exception, exc:
        if exc_args is not None:
            assert exc.args != exc_args, \
                        "%s raised %s with unexpected args: "\
                        "expected=%r, actual=%r"\
                        % (callsig, exc.__class__, exc_args, exc.args)
        if exc_pattern is not None:
            assert exc_pattern.search(str(exc)), \
                            "%s raised %s, but the exception "\
                            "does not match '%s': %r"\
                            % (callsig, exc.__class__, exc_pattern.pattern,
                               str(exc))
    except Exception, e:
        exc_info = sys.exc_info()
        print exc_info
        assert False, "%s raised an unexpected exception type: "\
                  "expected=%s, actual=%s (%s)"\
                  % (callsig, exception, exc_info[0], unicode(e))
    else:
        assert False, "%s did not raise %s" % (callsig, exception)


tests/unit/test_bolacha.py
#!/usr/bin/env python
# -*- coding: utf-8; -*-
#
# Copyright (C) 2009 Gabriel Falcão <gabriel@nacaolivre.org>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public
# License along with this program; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.
import types
from mox import Mox
from nose.tools import assert_equals
from utils import assert_raises

from bolacha import Bolacha, BOUNDARY
from httplib2 import Http

base_header = {'Content-type': 'application/x-www-form-urlencoded'}
def prepare_header(h):
    z = h.copy()
    z.update(base_header)
    return z

def test_creation_takes_class():
    msg = r'Bolacha takes a class or callable as parameter, got %r'
    assert_raises(TypeError,
                  Bolacha, 5,
                  exc_pattern=msg % 5)
    assert_raises(TypeError,
                  Bolacha, 'blabla',
                  exc_pattern=msg % 'blabla')

def test_instantiate_class_with_kwargs():
    mocker = Mox()

    klass_mock = mocker.CreateMockAnything()
    klass_mock(cache_path='/some/path', other_param=10)

    mocker.ReplayAll()
    Bolacha(klass_mock, cache_path='/some/path', other_param=10)
    mocker.VerifyAll()

def test_keeps_http_instance_after_instantiation():
    mocker = Mox()

    klass_mock = mocker.CreateMockAnything()
    instance_mock = mocker.CreateMockAnything()
    klass_mock().AndReturn(instance_mock)

    mocker.ReplayAll()
    b = Bolacha(klass_mock)
    assert_equals(b.http, instance_mock)
    mocker.VerifyAll()

def test_uses_httplib2_Http_as_default_http_class():
    b = Bolacha()
    assert isinstance(b.http, Http), 'Bolacha().http should be instance of ' \
           'httplib2.Http, got %s' % repr(b.http)

def test_is_persistent_by_default():
    b = Bolacha()
    assert b.persistent is True, 'Bolacha should be persistent by default'

def test_can_be_not_persistent():
    b = Bolacha(persistent=False)
    assert b.persistent is False, 'Bolacha should be persistent by default'

def test_is_persistent_by_default():
    b = Bolacha()
    assert b.persistent

def test_has_headers_attribute():
    b = Bolacha()
    assert hasattr(b, 'headers'), 'Bolacha should be "headers"'
    assert isinstance(b.headers, dict), \
           'Bolacha.headers should be a dict, got %s' % repr(b.headers)

def test_headers_are_instance_scoped():
    b1 = Bolacha()
    b2 = Bolacha()
    b1.headers['foo'] = 'foo foo'
    b2.headers['bar'] = 'bar bar'

    b3 = Bolacha()

    assert_equals(b1.headers, {'foo': 'foo foo'})
    assert_equals(b2.headers, {'bar': 'bar bar'})
    assert_equals(b3.headers, {})

def test_request_calls_http_request():
    mocker = Mox()

    klass_mock = mocker.CreateMockAnything()
    http_mock = mocker.CreateMockAnything()
    klass_mock().AndReturn(http_mock)

    response_headers = {}
    response_body = "should be my html content"

    http_mock.request('http://somewhere.com', 'GET', '', {}). \
        AndReturn((response_headers, response_body))

    mocker.ReplayAll()

    b = Bolacha(klass_mock)
    assert_equals(b.request('http://somewhere.com', 'GET'), (response_headers,
                                                             response_body))
    mocker.VerifyAll()

def test_request_fails_with_headers_non_dict():
    b = Bolacha()
    assert_raises(TypeError, b.request, 'http://somewhere', 'GET', headers=5,
                  exc_pattern=r'Bolacha.request, parameter headers must be ' \
                  'a dict or NoneType. Got 5')
    assert_raises(TypeError, b.request, 'http://somewhere', 'GET', headers='bla',
                  exc_pattern=r'Bolacha.request, parameter headers must be ' \
                  'a dict or NoneType. Got \'bla\'')

def test_request_fails_url_non_string():
    b = Bolacha()
    assert_raises(TypeError, b.request, None, None,
                  exc_pattern=r'Bolacha.request, parameter url must be ' \
                  'a string. Got None')
    assert_raises(TypeError, b.request, 99, None,
                  exc_pattern=r'Bolacha.request, parameter url must be ' \
                  'a string. Got 99')

def test_request_fails_method_non_string():
    b = Bolacha()
    assert_raises(TypeError, b.request, 'http://gnu', None,
                  exc_pattern=r'Bolacha.request, parameter method must be ' \
                  'a string. Got None')
    assert_raises(TypeError, b.request, 'http://gnu', 99,
                  exc_pattern=r'Bolacha.request, parameter method must be ' \
                  'a string. Got 99')

def test_request_fails_inexistent_method():
    b = Bolacha()
    assert_raises(TypeError, b.request, 'http://gnu', 'FOOBAR',
                  exc_pattern=r'Bolacha.request, parameter method must be ' \
                  'a valid HTTP method. Got FOOBAR. ' \
                  'Take a look at http://www.w3.org/Protocols/rfc2616/' \
                  'rfc2616-sec9.html to see valid method definitions')
    assert_raises(TypeError, b.request, 'http://gnu', 'GOOSFRABA',
                  exc_pattern=r'Bolacha.request, parameter method must be ' \
                  'a valid HTTP method. Got GOOSFRABA. ' \
                  'Take a look at http://www.w3.org/Protocols/rfc2616/' \
                  'rfc2616-sec9.html to see valid method definitions')

def test_request_with_body_dict():
    mocker = Mox()

    klass_mock = mocker.CreateMockAnything()
    http_mock = mocker.CreateMockAnything()
    klass_mock().AndReturn(http_mock)

    response_headers = {'header1': 'value of header'}
    response_body = {'param1': 'value1', 'foo': 'bar'}

    http_mock.request('http://somewhere.com', 'GET',
                      'foo=bar&param1=value1',
                      prepare_header(response_headers)). \
        AndReturn((response_headers, response_body))

    mocker.ReplayAll()

    b = Bolacha(klass_mock)
    got = b.request('http://somewhere.com', 'GET',
                    response_body,
                    prepare_header(response_headers))
    assert_equals(got, (response_headers,
                        response_body))
    mocker.VerifyAll()

def test_request_with_invalid_body():
    b = Bolacha()
    assert_raises(TypeError, b.request, 'http://gnu', 'GET',
                  body=['a list'],
                  exc_pattern=r'Bolacha.request, parameter body must be ' \
                  'a string or dict. Got .\'a list\'.')
    assert_raises(TypeError, b.request, 'http://gnu', 'GET',
                  body=5,
                  exc_pattern=r'Bolacha.request, parameter body must be ' \
                  'a string or dict. Got 5')

def test_request_keep_sending_last_cookies():
    mocker = Mox()

    http_mock = mocker.CreateMockAnything()

    response_headers1 = {'set-cookie': 5}
    response_headers2 = {'good': 10}
    response_headers3 = {'set-cookie': 20}

    request_headers1 = {}
    request_headers2 = {'Cookie': 5}
    request_headers3 = {'Cookie': 5}
    request_headers4 = {'Cookie': 20}

    # 1st request
    http_mock.request('http://somewhere.com', 'GET',
                      '', request_headers1). \
        AndReturn((response_headers1, ''))

    # 2nd request
    http_mock.request('http://somewhere.com', 'GET',
                      '', request_headers2). \
        AndReturn((response_headers2, ''))

    # 3rd request
    http_mock.request('http://somewhere.com', 'GET',
                      '', request_headers3). \
        AndReturn((response_headers3, ''))

    # 4th request
    http_mock.request('http://somewhere.com', 'GET',
                      '', request_headers4). \
        AndReturn(({}, ''))

    mocker.ReplayAll()
    bol = Bolacha()
    bol.http = http_mock
    for x in range(4):
        bol.request('http://somewhere.com', 'GET')

    mocker.VerifyAll()

def test_request_handle_cookies():
    mocker = Mox()

    http_mock = mocker.CreateMockAnything()
    request_headers1 = {}
    response_headers1 = {'set-cookie': 'Will log in'}

    request_headers2 = {'Cookie': 'Will log in'}
    response_headers2 = {'set-cookie': 'Already logged as root'}

    request_headers3 = {'Cookie': 'Already logged as root'}
    response_headers3 = {'set-cookie': 'Just logged out'}

    request_headers4 = {'Cookie': 'Just logged out'}
    response_headers4 = {'set-cookie': 'Will log in'}

    http_mock.request('http://somewhere.com', 'GET',
                      '', request_headers1). \
        AndReturn((response_headers1, ''))

    http_mock.request('http://somewhere.com', 'GET',
                      '', request_headers2). \
        AndReturn((response_headers2, ''))

    http_mock.request('http://somewhere.com', 'GET',
                      '', request_headers3). \
        AndReturn((response_headers3, ''))

    http_mock.request('http://somewhere.com', 'GET',
                      '', request_headers4). \
        AndReturn((response_headers4, ''))

    mocker.ReplayAll()
    bol = Bolacha()
    bol.http = http_mock
    for x in range(4):
        bol.request('http://somewhere.com', 'GET')

    mocker.VerifyAll()

def test_request_when_persistent():
    mocker = Mox()

    klass_mock = mocker.CreateMockAnything()
    http_mock = mocker.CreateMockAnything()
    klass_mock().AndReturn(http_mock)

    response_headers = {'connection': 'close'}

    http_mock.request('http://somewhere.com', 'GET', '', {}). \
        AndReturn((response_headers, ''))

    http_mock.request('http://somewhere.com', 'GET', '', {}).AndReturn(({}, ''))

    mocker.ReplayAll()

    b = Bolacha(klass_mock, persistent=True)
    b.request('http://somewhere.com', 'GET')
    b.request('http://somewhere.com', 'GET')
    mocker.VerifyAll()

def test_request_when_not_persistent():
    mocker = Mox()

    klass_mock = mocker.CreateMockAnything()
    http_mock = mocker.CreateMockAnything()
    klass_mock().AndReturn(http_mock)

    response_headers = {'connection': 'close', 'set-cookie': 'blabla'}
    request_headers = {'connection': 'close'}

    http_mock.request('http://somewhere.com', 'GET', '', {}). \
        AndReturn((response_headers, ''))

    http_mock.request('http://somewhere.com', 'GET', '', request_headers). \
        AndReturn((response_headers, ''))

    mocker.ReplayAll()

    b = Bolacha(klass_mock, persistent=False)
    b.request('http://somewhere.com', 'GET')
    b.request('http://somewhere.com', 'GET')
    mocker.VerifyAll()

def test_set_content_type_urlencoded_when_body_dict_and_none_was_given():
    mocker = Mox()

    klass_mock = mocker.CreateMockAnything()
    http_mock = mocker.CreateMockAnything()
    klass_mock().AndReturn(http_mock)

    request_headers = {'Content-type': 'application/x-www-form-urlencoded'}

    http_mock.request('http://somewhere.com', 'GET', '', request_headers). \
        AndReturn(({}, ''))

    mocker.ReplayAll()

    b = Bolacha(klass_mock)
    b.request('http://somewhere.com', 'GET', body={})
    mocker.VerifyAll()

def test_when_body_dict_and_content_type_is_specified():
    mocker = Mox()

    klass_mock = mocker.CreateMockAnything()
    http_mock = mocker.CreateMockAnything()
    klass_mock().AndReturn(http_mock)

    request_headers = {'Content-type': 'text/plain'}

    http_mock.request('http://somewhere.com', 'GET', '', request_headers). \
        AndReturn(({}, ''))

    mocker.ReplayAll()

    b = Bolacha(klass_mock)
    b.request('http://somewhere.com', 'GET', body={}, headers=request_headers)
    mocker.VerifyAll()

def test_post_shortcut():
    mocker = Mox()

    b = Bolacha()
    b.request = mocker.CreateMockAnything()

    b.request('host', 'POST', body={'name': 'foo', 'age': 30}, headers=None)
    mocker.ReplayAll()

    b.post('host', {'name': 'foo', 'age': 30})

    mocker.VerifyAll()

def test_get_shortcut():
    mocker = Mox()

    b = Bolacha()
    b.request = mocker.CreateMockAnything()

    b.request('host', 'GET', body={'name': 'foo', 'age': 30}, headers=None)
    mocker.ReplayAll()

    b.get('host', {'name': 'foo', 'age': 30})

    mocker.VerifyAll()

def test_put_shortcut():
    mocker = Mox()

    b = Bolacha()
    b.request = mocker.CreateMockAnything()

    b.request('host', 'PUT', body={'name': 'foo', 'age': 30}, headers=None)
    mocker.ReplayAll()

    b.put('host', {'name': 'foo', 'age': 30})

    mocker.VerifyAll()

def test_delete_shortcut():
    mocker = Mox()

    b = Bolacha()
    b.request = mocker.CreateMockAnything()

    b.request('host', 'DELETE', body={'name': 'foo', 'age': 30}, headers=None)
    mocker.ReplayAll()

    b.delete('host', {'name': 'foo', 'age': 30})

    mocker.VerifyAll()

def test_head_shortcut():
    mocker = Mox()

    b = Bolacha()
    b.request = mocker.CreateMockAnything()

    b.request('host', 'HEAD', body={'name': 'foo', 'age': 30}, headers=None)
    mocker.ReplayAll()

    b.head('host', {'name': 'foo', 'age': 30})

    mocker.VerifyAll()

def test_request_with_file_will_upload_multipart():
    mocker = Mox()

    klass_mock = mocker.CreateMockAnything()
    http_mock = mocker.CreateMockAnything()
    klass_mock().AndReturn(http_mock)

    class FileStub(object):
        name = '/path/to/file'
        def read(self):
            return 'FileStubContent'

    request_headers = {}
    body = {
        'name': 'John Doe',
        'picture': FileStub(),
    }

    expected_body = '--%(boundary)s\r\nContent-Disposition: form-data; ' \
                    'name="picture"; filename="file"\r\nContent-Type: ' \
                    'application/octet-stream\r\n\r\nFileStubContent\r\n' \
                    '--%(boundary)s\r\nContent-Disposition: form-data; ' \
                    'name="name"\r\n\r\nJohn Doe\r\n' \
                    '--%(boundary)s--\r\n' % {'boundary': BOUNDARY}
    expected_header = {'content-length': '291',
                       'Content-type': 'multipart/form-data; ' \
                       'boundary=%s' % BOUNDARY}

    http_mock.request('http://somewhere.com', 'POST',
                      expected_body, expected_header). \
        AndReturn(({}, ''))

    mocker.ReplayAll()

    b = Bolacha(klass_mock)
    b.request('http://somewhere.com', 'POST', body=body, headers=request_headers)
    mocker.VerifyAll()



tests/functional/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8; -*-
#
# Copyright (C) 2009 Gabriel Falcão <gabriel@nacaolivre.org>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public
# License along with this program; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.


bolacha.py
#!/usr/bin/env python
# -*- coding: utf-8; -*-
#
# Copyright (C) 2009 Gabriel Falcão <gabriel@nacaolivre.org>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public
# License along with this program; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.
from urllib import urlencode
from httplib2 import Http as HTTPClass

HTTP_METHODS = (
    'OPTIONS',
    'GET',
    'HEAD',
    'POST',
    'PUT',
    'DELETE',
    'TRACE',
    'CONNECT',
)

RFC_LOCATION = 'Take a look at http://www.w3.org/Protocols/rfc2616/' \
               'rfc2616-sec9.html to see valid method definitions'
class Bolacha(object):
    headers = None
    def __init__(self, http=None, persistent=True, **kw):
        if http is not None and not isinstance(http, type) and not callable(http):
            raise TypeError, 'Bolacha takes a class or callable as parameter, ' \
                  'got %s' % repr(http)

        if http is not None:
            self.http = http(**kw)
        else:
            self.http = HTTPClass(**kw)

        self.persistent = persistent
        self.headers = {}

    def request(self, url, method, body=None, headers=None):
        if not isinstance(url, basestring):
            raise TypeError, 'Bolacha.request, parameter url must be ' \
                  'a string. Got %s' % repr(url)

        if not isinstance(method, basestring):
            raise TypeError, 'Bolacha.request, parameter method must be ' \
                  'a string. Got %s' % repr(method)

        if method not in HTTP_METHODS:
            raise TypeError, 'Bolacha.request, parameter method must be ' \
                  'a valid HTTP method. Got %s. %s' % (method,
                                                        RFC_LOCATION)

        if body is None:
            body = ''

        if not isinstance(body, (basestring, dict)):
            raise TypeError, 'Bolacha.request, parameter body must be ' \
                  'a string or dict. Got %s.' % (repr(body))

        is_urlencoded = False
        if isinstance(body, dict):
            rbody = urlencode(body)
            is_urlencoded = True
        else:
            rbody = body

        if headers is None:
            headers = {}

        if not isinstance(headers, dict):
            raise TypeError, 'Bolacha.request, parameter headers must be ' \
                  'a dict or NoneType. Got %s' % repr(headers)

        rheaders = self.headers.copy()
        rheaders.update(headers)

        if self.persistent:
            if 'set-cookie' in self.headers:
                rheaders['Cookie'] = self.headers['set-cookie']

        if 'set-cookie' in rheaders:
            del rheaders['set-cookie']

        if self.persistent:
            if 'connection' in rheaders and rheaders['connection'] == 'close':
                del rheaders['connection']

        if is_urlencoded and not 'Content-type' in rheaders:
            rheaders['Content-type'] = 'application/x-www-form-urlencoded'

        response, content = self.http.request(url, method, rbody, rheaders)

        self.headers.update(response)

        return response, content


bolacha/multipart.py
#!/usr/bin/env python
# -*- coding: utf-8; -*-
#
# Copyright (C) 2009 Gabriel Falcão <gabriel@nacaolivre.org>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public
# License along with this program; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.

# some parts of the code below were took from Django Web Framework test client:
# http://code.djangoproject.com/browser/django/tags/releases/1.1/django/test/client.py
import types
from uuid import uuid4
from urllib import quote_plus, urlencode
from glob import glob
from os.path import basename
from mimetypes import guess_type

BOUNDARY = uuid4().hex

def is_file(obj):
    return hasattr(obj, 'read') and callable(obj.read)

def to_str(s, encoding='utf-8', strings_only=False, errors='strict'):
    """ took from django smart_str """

    if strings_only and isinstance(s, (types.NoneType, int)):
        return s

    if not isinstance(s, basestring):
        try:
            return str(s)
        except UnicodeEncodeError:
            if isinstance(s, Exception):
                # An Exception subclass containing non-ASCII data that doesn't
                # know how to print itself properly. We shouldn't raise a
                # further exception.
                return ' '.join([smart_str(arg, encoding, strings_only,
                        errors) for arg in s])
            return unicode(s).encode(encoding, errors)
    elif isinstance(s, unicode):
        return s.encode(encoding, errors)
    elif s and encoding != 'utf-8':
        return s.decode('utf-8', errors).encode(encoding, errors)
    else:
        return s

def encode_multipart(boundary, data):
    lines = []

    for key, value in data.items():
        if is_file(value):
            lines.extend(encode_file(boundary, key, value))
        else:
            if is_file(value):
                lines.extend(encode_file(boundary, key, value))
            else:
                lines.extend([
                    '--' + boundary,
                    'Content-Disposition: form-data; name="%s"' % to_str(key),
                    '',
                    to_str(value)
                ])

    lines.extend([
        '--' + boundary + '--',
        '',
    ])
    return '\r\n'.join(lines)

def guess_mime(path):
    mime, x = guess_type(path)
    return mime or 'application/octet-stream'

def encode_file(boundary, key, file):
    return [
        '--' + boundary,
        'Content-Disposition: form-data; name="%s"; filename="%s"' \
            % (to_str(key), to_str(basename(file.name))),
        'Content-Type: %s' % guess_mime(file.name),
        '',
        to_str(file.read())
    ]


bolacha/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8; -*-
#
# Copyright (C) 2009 Gabriel Falcão <gabriel@nacaolivre.org>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public
# License along with this program; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.
from httplib2 import Http as HTTPClass
from bolacha.multipart import BOUNDARY
from bolacha.multipart import encode_multipart
from bolacha.multipart import urlencode
from bolacha.multipart import is_file

__version__ = '0.2.1'
__release__ = 'Chewy'

HTTP_METHODS = (
    'OPTIONS',
    'GET',
    'HEAD',
    'POST',
    'PUT',
    'DELETE',
    'TRACE',
    'CONNECT',
)

RFC_LOCATION = 'Take a look at http://www.w3.org/Protocols/rfc2616/' \
               'rfc2616-sec9.html to see valid method definitions'

class Bolacha(object):
    headers = None
    def __init__(self, http=None, persistent=True, **kw):
        if http is not None and not isinstance(http, type) and not callable(http):
            raise TypeError, 'Bolacha takes a class or callable as parameter, ' \
                  'got %s' % repr(http)

        if http is not None:
            self.http = http(**kw)
        else:
            self.http = HTTPClass(**kw)

        self.persistent = persistent
        self.headers = {}

    def post(self, url, body=None, headers=None):
        return self.request(url, 'POST', body=body, headers=headers)

    def get(self, url, body=None, headers=None):
        return self.request(url, 'GET', body=body, headers=headers)

    def put(self, url, body=None, headers=None):
        return self.request(url, 'PUT', body=body, headers=headers)

    def delete(self, url, body=None, headers=None):
        return self.request(url, 'DELETE', body=body, headers=headers)

    def head(self, url, body=None, headers=None):
        return self.request(url, 'HEAD', body=body, headers=headers)

    def request(self, url, method, body=None, headers=None):
        if not isinstance(url, basestring):
            raise TypeError, 'Bolacha.request, parameter url must be ' \
                  'a string. Got %s' % repr(url)

        if not isinstance(method, basestring):
            raise TypeError, 'Bolacha.request, parameter method must be ' \
                  'a string. Got %s' % repr(method)

        if method not in HTTP_METHODS:
            raise TypeError, 'Bolacha.request, parameter method must be ' \
                  'a valid HTTP method. Got %s. %s' % (method,
                                                        RFC_LOCATION)

        if body is None:
            body = ''

        if not isinstance(body, (basestring, dict)):
            raise TypeError, 'Bolacha.request, parameter body must be ' \
                  'a string or dict. Got %s.' % (repr(body))

        is_urlencoded = False

        body_has_file = isinstance(body, dict) and any([is_file(fobj)
                                                        for fobj in body.values()])

        if isinstance(body, dict):
            if body_has_file:
                rbody = encode_multipart(BOUNDARY, body)
            else:
                rbody = urlencode(body)
                is_urlencoded = True
        else:
            rbody = body

        if headers is None:
            headers = {}

        if not isinstance(headers, dict):
            raise TypeError, 'Bolacha.request, parameter headers must be ' \
                  'a dict or NoneType. Got %s' % repr(headers)

        rheaders = self.headers.copy()
        rheaders.update(headers)

        if self.persistent:
            if 'set-cookie' in self.headers:
                rheaders['Cookie'] = self.headers['set-cookie']

        if 'set-cookie' in rheaders:
            del rheaders['set-cookie']

        if is_urlencoded and not 'Content-type' in rheaders:
            rheaders['Content-type'] = 'application/x-www-form-urlencoded'
        elif body_has_file:
            rheaders['Content-type'] = 'multipart/form-data; boundary=%s' % BOUNDARY
            rheaders['content-length'] = '%d' % len(rbody)

        response, content = self.http.request(url, method, rbody, rheaders)

        if self.persistent and 'set-cookie' in response:
            self.headers['set-cookie'] = response['set-cookie']

        if not self.persistent:
            if 'connection' in response:
                self.headers['connection'] = response['connection']

        return response, content
